transaction/CycleDetector.cpp

```diff
@@ -0,0 +1,122 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#include "transaction/CycleDetector.hpp"
+
+#include <cstdint>
+#include <memory>
+#include <stack>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "transaction/DirectedGraph.hpp"
+#include "transaction/StronglyConnectedComponents.hpp"
+
+namespace quickstep {
+namespace transaction {
+
+CycleDetector::CycleDetector(DirectedGraph *wait_for_graph)
+    : wait_for_graph_(wait_for_graph),
+      strongly_connected_components_(
+          std::make_unique<StronglyConnectedComponents>(*wait_for_graph)) {
+}
+
+std::vector<DirectedGraph::node_id> CycleDetector::breakCycle() {
+  std::vector<DirectedGraph::node_id> nodes_to_kill;
+  const std::unordered_map<std::uint64_t, std::vector<DirectedGraph::node_id>>
+    component_mapping = strongly_connected_components_->getComponentMapping();
+  for (const std::pair<std::uint64_t, std::vector<DirectedGraph::node_id>>
+           &entry : component_mapping) {
+    // One node means no cycle.
+    if (entry.second.size() == 1) {
+      continue;
+    }
+    const std::vector<DirectedGraph::node_id> nodes
+        = breakComponent(entry.second);
+      for (const DirectedGraph::node_id &node : nodes) {
+        nodes_to_kill.push_back(node);
+      }
+  }
+  return nodes_to_kill;
+}
+
+std::vector<DirectedGraph::node_id>
+CycleDetector::breakComponent(const std::vector<DirectedGraph::node_id> &nodes) {
+  std::vector<DirectedGraph::node_id> targets;
+  // Convert it to set to ensure defensively that the elements are unique.
+  std::unordered_set<DirectedGraph::node_id> nodes_set(nodes.begin(), nodes.end());
+
+  while (true) {
+    const bool has_cycle = hasCycleWithin(nodes_set);
+    if (!has_cycle) {
```

Comment:
Since `has_cycle` does not use in anywhere else, how about the following refactor:

```
  if (!hasCycleWithin(nodes_set)) {
    break;
  }
```

Also, regarding the method name, I prefer `hasCycle` instead of `hasCycleWithin`.